designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala

```diff
@@ -0,0 +1,160 @@
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+import pl.touk.nussknacker.ui.BadRequestError
+
+class CustomActionValidator(val allowedActions: List[CustomActionDefinition]) {
+
+  def validateCustomActionParams(
+      request: CustomActionRequest
+  ): Either[CustomActionValidationError, CustomActionValidationResult] = {
+
+    val checkedCustomAction =
+      allowedActions
+        .find(_.name == request.actionName)
+        .toRight(CustomActionNonExistingError("Couldn't find this action: " + request.actionName.value))
+
+    checkedCustomAction match {
+      case Left(notFoundAction) => Left(notFoundAction)
+
+      case Right(foundAction) => {
+        implicit val nodeId: NodeId = NodeId(foundAction.name.value)
+        val customActionParams      = foundAction.parameters
+        val requestParamsMap        = getRequestParamsMap(request, customActionParams)
+
+        val validated = validateParams(requestParamsMap, customActionParams)
+        getValidationResult(validated)
+      }
+    }
+  }
+
+  private def getRequestParamsMap(request: CustomActionRequest, customActionParams: List[CustomActionParameter]) = {
+    (customActionParams.nonEmpty, request.params) match {
+      case (true, Some(paramsMap)) =>
+        if (paramsMap.keys.size != customActionParams.size) {
+          Left(
+            MismatchedParamsError(
+              s"Validation requires different count of custom action parameters than provided in request for: ${request.actionName}"
+            )
+          )
+        } else { Right(paramsMap) }
+      case (true, None) =>
+        Left(MismatchedParamsError(s"Missing required params for action: ${request.actionName}"))
+      case (false, Some(_)) =>
+        Left(MismatchedParamsError(s"Params found for no params action: ${request.actionName}"))
+      case _ => Right(Map.empty[String, String])
+    }
+  }
+
+  private def validateParams(
+      requestParamsMap: Either[CustomActionValidationError, Map[String, String]],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): Either[CustomActionValidationError, Map[String, List[PartSubGraphCompilationError]]] = {
+    val checkedParamsMap = checkForMissingKeys(requestParamsMap, customActionParams)
+
+    checkedParamsMap.map { m =>
+      m.map { case (k, v) =>
+        (
+          k,
+          customActionParams.find(_.name == k) match {
+            case Some(param) =>
+              param.validators
+                .getOrElse(Nil)
+                .map {
+                  _.isValid(
+                    paramName = ParameterName(k),
+                    expression = Expression.spel("None"),
+                    value = Some(v),
+                    label = None
+                  )
+                }
+                .collect { case Invalid(i) => i }
+            case None =>
+              throw new IllegalStateException
+          }
+        )
+      }
+    }
+  }
+
+  private def checkForMissingKeys(
```

Comment:
See similar validation in ScenarioPropertiesValidator

Reply:
removed